Mathlib/Analysis/NormedSpace/ProdLp.lean

```diff
@@ -0,0 +1,743 @@
+  rfl
+
+@[simp]
+theorem zero_snd : (0 : WithLp p (α × β)).snd = 0 :=
+  rfl
+
+@[simp]
+theorem add_fst : (x + y).fst = x.fst + y.fst :=
+  rfl
+
+@[simp]
+theorem add_snd : (x + y).snd = x.snd + y.snd :=
+  rfl
+
+@[simp]
+theorem sub_fst : (x - y).fst = x.fst - y.fst :=
+  rfl
+
+@[simp]
+theorem sub_snd : (x - y).snd = x.snd - y.snd :=
+  rfl
+
+@[simp]
+theorem smul_fst : (c • x).fst = c • x.fst :=
+  rfl
+
+@[simp]
+theorem smul_snd : (c • x).snd = c • x.snd :=
+  rfl
+
+@[simp]
+theorem neg_fst : (-x).fst = -x.fst :=
+  rfl
+
+@[simp]
+theorem neg_snd : (-x).snd = -x.snd :=
+  rfl
+
+end algebra
+
+section Equiv
+
+/-- The canonical map `ProdLp.equiv` between `ProdLp ∞ β` and `Π i, β i` as a linear isometric
+equivalence. -/
+def prodEquivₗᵢ : WithLp ∞ (α × β) ≃ₗᵢ[𝕜] α × β :=
+  { WithLp.equiv ∞ (α × β) with
+    map_add' := fun f g => rfl
+    map_smul' := fun c f => rfl
+    norm_map' := fun f => by simp [Norm.norm] }
+
+variable (x y : WithLp p (α × β)) (x' y' : α × β) (c : 𝕜)
+
+end Equiv
+
+section Single
+
+variable (p α β)
+
+@[simp]
+theorem nnnorm_equiv_symm_fst [hp : Fact (1 ≤ p)] (x : α) :
+    ‖(WithLp.equiv p (α × β)).symm (x, 0)‖₊ = ‖x‖₊ := by
+  induction p using ENNReal.recTopCoe generalizing hp with
+  | top =>
+    simp [prod_nnnorm_eq_sup]
+  | coe p =>
+    have hp0 : (p : ℝ) ≠ 0 := by
+      exact_mod_cast (zero_lt_one.trans_le <| Fact.out (p := 1 ≤ (p : ℝ≥0∞))).ne'
+    simp [prod_nnnorm_eq_add, NNReal.zero_rpow hp0, ← NNReal.rpow_mul, mul_inv_cancel hp0]
+
+@[simp]
+theorem nnnorm_equiv_symm_snd [hp : Fact (1 ≤ p)] (y : β) :
+    ‖(WithLp.equiv p (α × β)).symm (0, y)‖₊ = ‖y‖₊ := by
+  induction p using ENNReal.recTopCoe generalizing hp with
+  | top =>
+    simp [prod_nnnorm_eq_sup]
+  | coe p =>
+    have hp0 : (p : ℝ) ≠ 0 := by
+      exact_mod_cast (zero_lt_one.trans_le <| Fact.out (p := 1 ≤ (p : ℝ≥0∞))).ne'
+    simp [prod_nnnorm_eq_add, NNReal.zero_rpow hp0, ← NNReal.rpow_mul, mul_inv_cancel hp0]
+
+@[simp]
+theorem norm_equiv_symm_fst (x : α) : ‖(WithLp.equiv p (α × β)).symm (x, 0)‖ = ‖x‖ :=
+  congr_arg ((↑) : ℝ≥0 → ℝ) <| nnnorm_equiv_symm_fst p α β x
+
+@[simp]
+theorem norm_equiv_symm_snd (y : β) : ‖(WithLp.equiv p (α × β)).symm (0, y)‖ = ‖y‖ :=
+  congr_arg ((↑) : ℝ≥0 → ℝ) <| nnnorm_equiv_symm_snd p α β y
+
+@[simp]
+theorem nndist_equiv_symm_single_fst (x₁ x₂ : α) :
+    nndist ((WithLp.equiv p (α × β)).symm (x₁, 0)) ((WithLp.equiv p (α × β)).symm (x₂, 0)) =
+      nndist x₁ x₂ := by
+  rw [nndist_eq_nnnorm, nndist_eq_nnnorm, ← WithLp.equiv_symm_sub, Prod.mk_sub_mk, sub_zero,
+    nnnorm_equiv_symm_fst]
+
+@[simp]
+theorem nndist_equiv_symm_single_snd (y₁ y₂ : β) :
+    nndist ((WithLp.equiv p (α × β)).symm (0, y₁)) ((WithLp.equiv p (α × β)).symm (0, y₂)) =
+      nndist y₁ y₂ := by
+  rw [nndist_eq_nnnorm, nndist_eq_nnnorm, ← WithLp.equiv_symm_sub, Prod.mk_sub_mk, sub_zero,
+    nnnorm_equiv_symm_snd]
+
+@[simp]
+theorem dist_equiv_symm_single_fst (x₁ x₂ : α) :
```

Comment:
These `dist` and `edist` results could be stated earlier with weaker type classes

Reply:
I've removed the assumption on normed spaces, but due to the structure of the proofs we still require a more assumptions than really necessary (otherwise we do not get the `nnnorm`).